### ISSUE TITLE
New feature: author pages

### ISSUE THREAD

**Issue by librarianette** [braxtonhall/library-thing#66]
TYSM for all your work so far!!!!! 

one thing that Natalie brought up that would be great is some form of making sure that an author's identity tags stay consistent across different works of theirs. If it helps, each author has their own LibraryThing page, which stores data like their birth and death date, nationality, sex, education, etc. Unfortunately, it doesn't store the things that we want to track (besides nationality), but it might be useful?

Let me know if you need more info!

- Nola

**Comment by braxtonhall** [comment]
OK so here's the current vision:

# Version 1

Author data is stored remotely in a Google Sheet
| UUID   | Name          | Tags                               |
|--------|---------------|------------------------------------|
| 123456 | Jane Doe      | Canadian author, Femme author      |

Where
- UUID is the LibraryThing internal ID.
- Name is the LibraryThing human readable name.
- Tags is the list of author tags. Invariant: all tags include the word "author"

When visiting an author page, there is a **entry field**, and a **save button**

Here is the pseudocode for the author page
```python
def on_page_render():
    author_id, author_name = document.get_author_details()
    ensure_author(author_id, author_name)
    entry_field.text = get_tags(author_id)

def ensure_author(author_id, author_name):
    google_doc = get_google_doc()
    if not author_id in google_doc:
        google_doc.create_row(author_id, author_name, None)

def get_tags(author_id):
    google_doc = get_google_doc()
    if author_id in google_doc:
        return google_doc.get_tags(author_id)
    else:
        return None

def on_save_button_click():
    author_id, = document.get_author_details()
    tags = get_tags(author_id)
    for book in documents.get_books:
        update_tags_for_root_book(author_id, book, tags)

def update_tags_for_root_book(author_id, root_book, tags):
    books = [root_book] + document.get_other_editions(root_book)
    for book in books:
        update_tags_for_book(book, tags)

def update_tags_for_book(author_id, book, author_tags):
    book_tags = book.author_tags
    maybe_delete_tags = book_tags - author_tags
    other_author_tags = book.authors.filter(_ not equals author_id).flat(get_tags)
    delete_tags = maybe_delete_tags.filter(_ not in other_author_tags)
    book.author_tags = (book_tags + author_tags) - delete_tags

```

# Version 2
In some future update, it would be nice to add tags directly to the book page (say the author does not yet exist)

The book page would have a new button called "push" that adds all new tags to the author in the google doc, and as bonus performs the sync operation

**Comment by braxtonhall** [comment]
It would probably also make sense that you can, from the author page, pull up tags from all of the books. This would help with old entries from before this update goes live.

OK this issue needs to get decomposed.
- [ ] Google Docs API
  - [ ] Google Docs API key register (a pop up asks for an API key which is saved to localStorage. it is shared between volunteers like a password)
- [ ] Author info API
- [ ] Lastly, this feature

**Comment by adirar111** [comment]
### st8 of things

I was planning on using [Google's client side library for sheets](https://developers.google.com/sheets/api/quickstart/js#python-3.x), but this won't work because it requires having our extension dynamically load a script into the page (this isn't allowed with manifest v3, and we can't go back to manifest v2 because it's getting deprecated).

Manifest V3 documentation suggests that if you need to use a library, you should [bundle it with your extension](https://developer.chrome.com/docs/extensions/mv3/intro/mv3-migration/#code-execution). Attempting this exposed the `gapi` object for use, but when trying to initialize it via `gapi.load(...)`, it tried to dynamically load more scripts into the page (which were blocked). So this doesn't seem like a viable approach.

There's a node npm package, but it uses node APIs so we can't use it on the client (also webpack was failing loudly).

Other extension developers noticed this too:
* https://stackoverflow.com/questions/65625854/how-to-integrate-gapi-in-chrome-extensions-manifest-v3
* https://bugs.chromium.org/p/chromium/issues/detail?id=1164452

To put an end to any ambiguity on future support, the Google API team has [updated their documentation to reflect that the client side library will no longer be supported in chrome extensions](https://github.com/google/google-api-javascript-client/blob/master/docs/start.md#supported-environments).

### REST API approach
Sheets does expose a REST API that can be used to do the same things that the library does (though this is a little more involved). Trying to hit the endpoint that reads data from a sheet works and the only authorization needed is an API key from the Google Cloud project (yay!)

If you want to do anything more interesting like append/update rows, you'll need to have users go through their OAuth flow. There's a [client-side example for this provided in the OAuth docs](https://developers.google.com/identity/protocols/oauth2/javascript-implicit-flow#example) that's achieved by making a request to the OAuth endpoint and saving the returned Bearer token to `localStorage`. To authorize users, it redirects them to Google's permission page, they sign in, it approves, it redirects them back to the page they were on with a token (I have an implementation of this that half-works, but LibraryThing breaks when the authroization token gets appended to the URL so still need to figure that out).

The fact that we have to deal with OAuth at all is in and of itself annoying, but ALSO, (from what I understand) it MIGHT mean getting our app/project approved/verified by Google (which apparently can take a while). Without app verification, they cap the number of users that can do OAuth per project to 100 (which I doubt will be a problem for our scale, but still).

Also to be fair, I think we would've had to deal with OAuth with the client library anyways, but it would have provided a nice layer of abstraction.

TLDR: we'll be able to read/ write to a google sheet, but it's gonna be kinda annoying and things might be very suceptible to breaking as we figure out how to navigate all the Google Cloud-isms.

**Comment by braxtonhall** [comment]
Would it be INSANE to query the sheet to figure out where each column is? Or do we just say that the sheet can never change the order of the columns?

**Comment by adirar111** [comment]
> Would it be INSANE to query the sheet to figure out where each column is? Or do we just say that the sheet can never change the order of the columns?

rn it's kinda insane lmao -- if an invariant for this is that all tags have an "author" suffix, it seems reasonable that another invariant would be column order. that being said, idk how big/ complex this data store will be (esp when we do nested tags), so it might be worth looking into later?

**Comment by braxtonhall** [comment]
This feature is basically done. Don't want to move forward with the "Version 2" stuff, as better semantics are described in #91 and #94
